### ISSUE TITLE
Improve download packages for CentOS

### ISSUE THREAD

**Issue by slaperche-scality** [scality/metalk8s#1720]
**Component**:

build

**Why this is needed**:

In order to have smaller ISO/faster build time (and also because it's always better to pull only what's needed), we should rework how we download RPM packages for CentOS.

From my firsts tests, it seems that we could reduce the number of packages from 513 to 215, saving ~250M which would nicely compensate the size increase due to Ubuntu (~100M).

**What should be done**:

Use the approach that was used for Ubuntu (see https://github.com/scality/metalk8s/pull/1670), : write a Python script (even easier here since Yum is written in Python) to resolve the dependencies ourself and fetch only what's needed.

**Implementation proposal**:

Follow the "algorithm" used for Ubuntu.
For each packages we explicitly install, we resolve its dependencies (and the dependencies of its dependencies, all the way down…) and add all of this into a list (with no duplicate).
Then, we download each package from the list.

**Test plan**:

Build works as before and the generated ISO can be used to deploy a MetalK8s cluster.

**Comment by slaperche-scality** [comment]
I already have a working PoC for this, just need to check if the resulting ISO is working fine!

**Comment by jbertran** [comment]
This could also be the point where we give up formatting package names as strings entirely, i.e. remove some part of #1702 that is here in lieu of better management because of this very ticket.

Isofunctionality with respect to `versions.json` creation should be considered.

**Comment by slaperche-scality** [comment]
yeah, we could serialize the the packages info as JSON and mount that into the docker (or pass it somehow).

Unfortunately, we can't `import versions` directly (at least for CentOS, because using Yum means using Python2 and the file is clearly not compatible with py2 :stuck_out_tongue: ), so mouting `versions.py` is not an option.

To be defined if we do it in this PR or in a followup one.

**Comment by gdemonet** [comment]
Serialized as JSON is roughly what's done for Salt already, shouldn't be too much of hassle. Follow-up seems reasonable as well.

Could we also make sure the script now fails if some requested packages are not available (that's not the case with `yum-downloader` IIRC)?

**Comment by slaperche-scality** [comment]
> Could we also make sure the script now fails if some requested packages are not available (that's not the case with yum-downloader IIRC)?

To be confirmed, but I guess the API already raise an exception if it can't download (but I'll test it)

**Comment by slaperche-scality** [comment]
Hum, I'm actually encountering an issue on that one (see [here](https://eve.devsca.com/github/scality/metalk8s/#/builders/1/builds/11253)):

```
Error: Package: systemd-sysv-219-57.el7_5.3.x86_64 (installed)
           Requires: systemd = 219-57.el7_5.3
           Removing: systemd-219-57.el7_5.3.x86_64 (installed)
               systemd = 219-57.el7_5.3
           Updated By: systemd-219-67.el7_7.1.x86_64 (metalk8s-updates)
               systemd = 219-67.el7_7.1
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```

From what I understand:
- we want to install `runc`
- `runc` depends on `systemd`, so we package the most recent available on the build container (the `219-67` one)
- then, we deploy MetalK8s and tries to install `runc`
- there is already a `systemd` available on the OS (the `219-57`) and `yum` in its eternal wisdom tries to update it because it sees our newest version (the `219-67`).
- Problem is, `systemd-sysv-219-57` is also installed on the system and it depends on `systemd-219-57`, and by upgrading `systemd` we break `systemd-sysv` (because we didn't ship `systemd-sysv-219-67` since none of our packages depends on it)

I guess that's why we currently ship 250M of useless packages: just in case…

Now wondering if we have a way to keep downloading a minimal set of package while keeping a working ISO.

**Comment by NicolasT** [comment]
That's *exactly* why we ship `@core` and `@base`... There may be other cases (but definitely the case for `systemd-*` packages) where indeed we (maybe transitively) depend on a package which is a (version-constrained) dependency of another package we *don't* depend on, and then indeed, breakage.

So
- either we ship `@core` and `@base` as today, which shouldn't cause an issue unless upstream is broken
- we only ship our deps and their deps that are *not* in `@core` or `@base`, and cross fingers there are repositories configured on the system that permit to install things from these groups (which is, from experience, not always the case)
- 'blacklist' a bunch of packages from `@core` and `@base` and don't include them in what we mirror from these groups, though that seems very feeble.

**Comment by jbertran** [comment]
If the issue is localized to `systemd`, there appears to be an exclude/lock mechanism in `yum` that would allow us to prevent it from updating.

However that means that we're playing catchup from this point on, which is not ideal.

**Comment by NicolasT** [comment]
Yeah, `systemd-foo-VERSION-RELEASE` hard-depends on `systemd-VERSION-RELEASE`. Same goes for the Python bindings and such.

So, maybe blacklisting `systemd-*` from what we mirror in transitive deps could be sufficient? :man_shrugging: However, if at some point e.g. `runc` were to depend on `systemd >= VERSION` from `updates` and we don't ship it to an environment which hasn't been updated, we'd be stuck again... However, that's an 'if ever.

Actually, we could have a CI step which sets up a YUM environment (through the API, even) which *only* includes the release repo, not *updates* and others, then validates all packages we *want* to install during install can be fully resolved from said repo + the offline repos we provide. In that case the 'what if' scenario above would be caught.

Remains the issue of release upgrades (target system running CentOS 7.5, our CI test above running at a time 7.7 has been released, passing, but our packages having a hard req on `systemd-*` from 7.7, which is at that point not available in the 7.5 environment... So OS-release-support wise we can only claim compat with 'latest' releases, unless we somehow target full mirrors, or at least have `repodata` mirrors someplace).

**Comment by slaperche-scality** [comment]
Hum, IIUC the safest path is to keep shipping `base` and `core` (and I guess we will have to do something similar for Ubuntu…).

Still worth to rewrite in Python though (easier to maintain/evolve, can use `versions.py` almost directly, …), but we won't have the nice space saving (at least not a noticeable one).

**Comment by slaperche-scality** [comment]
For the record (in case someone wants to take over this if I don't), here is what I came up with:

```python
#!/usr/bin/env python2
# coding: utf-8

import errno
import os
import shutil
import sys

import yum


# Helpers {{{


def get_package(sack, package_name):
    # type: (yum.packageSack.MetaSack, str) -> yum.packages.YumAvailablePackage
    """Return the specified package."""
    name, _, version = package_name.partition('=')
    if version:
        package = sack.searchNevra(name=name, ver=version)
    else:
        package = sack.returnNewestByName(name=name)
    assert len(package) == 1
    return package[0]


# }}}

def get_package_deps(
    sack,              # type: yum.packageSack.MetaSack
    root_package_name  # type: str
):
    # type: (...) -> Dict[str, yum.packages.YumAvailablePackage]
    """Return the given package and all of its dependencies.

    This function works recursively, traversing the whole dependency tree
    starting from `root_package`.
    """
    dependencies = {}
    stack = [get_package(sack, root_package_name)]
    while stack:
        package = stack.pop()
        # Skip already added dependency.
        if package.ui_nevra in dependencies:
            continue
        dependencies[package.ui_nevra] = package
        for requisite in package.requires:
            candidates = sack.getProvides(*requisite)
            if not candidates:  # Can happen because we exclude some arch.
                continue
            # Take the most recent package that satisfy the dependency.
            stack.append(sorted(candidates.keys())[-1])
    return dependencies


def download_package(package):
    # type: (yum.packages.YumAvailablePackage) -> str
    """Download the RPM corresponding to `package`."""
    print('Downloading package {}'.format(package.ui_nevra))
    pkg_path = package.repo.getPackage(package)
    # Copy the package into the expected location.
    filepath = os.path.join(
        '/repositories',
        'metalk8s-{}-el7'.format(package.ui_from_repo),
        os.path.basename(pkg_path)
    )
    shutil.move(pkg_path, filepath)
    return filepath

def main(packages, env):
    # type: (Sequence[str], Mapping[str, str]) -> None
    """Download the packages specified on the command-line."""
    yb = yum.YumBase()
    yb.setCacheDir()
    # Yeah, this looks weird but here is the doc:
    #     exclude incompatible arches. archlist is a list of compatible arches
    # So yes, here I'm excluding non-x86_64 stuff…
    yb.pkgSack.excludeArchs(archlist=['x86_64', 'noarch'])
    pkg = get_package(yb.pkgSack, packages[0])
    to_download = {}
    for package in packages:
        deps = get_package_deps(yb.pkgSack, package)
        to_download.update(deps)
    for pkg in to_download.values():
        filepath = download_package(pkg)
        os.chown(filepath, int(env['TARGET_UID']), int(env['TARGET_GID']))


if __name__ == '__main__':
    main(sys.argv[1:], os.environ)
```

The download of the whole `core`/`base` we've discussed hasn't been implemented yet.

Edit: useful resources:
- https://linuxhint.com/yum_centos_python/
- http://yum.baseurl.org/api/yum/index-2.html
